File: typescript/0643-maximum-average-subarray-i.ts
/**
 * 643 - Maximum Average Subarray I
 * Sliding window | Fixed size window
 * Time: O(n) | Space: O(1)
 */

function findMaxAverage(nums: number[], k: number): number {
  let curr = 0;

  for (let i = 0; i < k; i++) {
    curr += nums[i];
  }

  let res = curr / k;

  for (let i = k; i < nums.length; i++) {
    curr += nums[i] - nums[i - k];
    res = Math.max(res, curr / k);
  }

  return res;
}

File: packages/python/src/leetcode/by-problem/0643-maximum-average-subarray-i.py
"""
643 - Maximum Average Subarray I [Easy]
Sliding Window | Fixed Window Size | Array
Time: O(n) | Space: O(1)

You are given an integer array nums consisting of n elements, and an integer k.

Find a contiguous subarray whose length is equal to k that has the maximum average value and return this value.
Any answer with a calculation error less than 10^-5 will be accepted.
"""

from typing import List


class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        curr = 0

        for i in range(k):
            curr += nums[i]

        res = curr

        for i in range(k, len(nums)):
            curr += nums[i] - nums[i - k]
            res = max(res, curr)

        return res / k
